Post Content Block: Render readonly content as blocks to preserve block supports styles

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->
This is an exploratory PR to come up with a shorter-term fix for #35376. By rendering the post content block as a live preview within the editor, we resolve the following issues:

* Correctly render the social links icons in this context (resolves: #35376).
* Ensure Duotone filters are rendered for image blocks within post content (resolves part of https://github.com/WordPress/gutenberg/issues/35331).
* Ensure custom link colors are rendered within the post content block.

We achieve the live block preview by doing the following:

* Add a `useDisabled` hook to provide a hook-based approach that achieves most of the functionality of the `Disabled` component, without adding a wrapper element to the DOM.
* Add a `useBlockPreview` hook, that behaves similarly to the `useInnerBlocksProps` hook, by allowing the consuming component to add the behaviour of a block preview to a wrapper element. This turns a component into a block preview, without adding an additional wrapper DOM element.
* The `useBlockPreview` hook uses `useDisabled` internally. The Post Content block is then updated to use `useBlockPreview` for the live block preview.

## To-do

* [x] The appender is rendered within the BlockPreview, but cannot be clicked on (fixed)
* [x] Because the BlockPreview renders the blocks in a Disabled state, the Social Links icons are greyed out (fixed — added a CSS rule)
* [x] Placeholders are rendered (fixed — added a CSS rule) 
* Clicking between two Post Content blocks results in the vertical scroll of the page going to the wrong place — this appears to be behaving similarly on trunk — I believe it's specific to clicking between two similar blocks within a Query loop block within the editor, so might be something for us to look into separately from this PR. Update: I have a fix for the vertical scroll issue in #35945 and have raised the issue with clicking between blocks within the Post Template block in #35944

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

* Activate an FSE theme like TT1-Blocks
* Publish a post containing the following:
  - An image block using the Duotone filter
  - Social links blocks
  - A block with a custom link color (e.g. Paragraph block, and insert an inline link)
* Open up the site editor, and if a Query loop block does not yet exist in your template, insert the block, and within it, use the Post Content block
* The post you previously published should be rendered correctly in the editor

## Screenshots <!-- if applicable -->

Note the following includes blocks that use Layout spacing, Duotone, and a custom link color (block supports that we need to ensure render correctly within the read only mode of the Post Content block):

| Before | After |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/14988353/138647208-7b14f6df-922f-40c9-89fe-cc8c953e62b5.png) | ![image](https://user-images.githubusercontent.com/14988353/138639917-a8f8963d-c4e9-4f68-b9cf-2e301fc38baa.png) |

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
Post Content Block: try rendering read only content as blocks to preserve block supports styles

andrewserong committed:
Prevent post content inner from being tabbable, fix disabled styles, hide placeholders

ramonjd commented:
Here's what I'm seeing:

<img width="930" alt="Screen Shot 2021-10-26 at 11 42 42 am" src="https://user-images.githubusercontent.com/6458278/138790182-cf2ed285-b6c8-4c9f-b025-1f7eb451a46f.png">

✅ The duotone filter, social icons links and custom paragraph link colors are rendering correctly.
✅ I can't tab or select post content, or interact with it in any way.
✅ Multiple query loops in the same templates behave similarly
✅ Blocks such as Columns behave correctly in different viewport widths, i.e., they collapse in mobile preview

I also inserted a bunch of other blocks just to see. All looked as expected.

It'd be interesting to get the opinions of Site Editor experts, but based on [these comments](https://github.com/WordPress/gutenberg/pull/35450#issuecomment-951391011) I think you've landed on a very nice workaround 🥇

youknowriad commented on packages/block-library/src/post-content/editor.scss:
```diff
@@ -0,0 +1,20 @@
+		.block-editor-block-list__block[data-empty="true"] {
+			display: none;
+		}
+	}
```
Are the styles here specific to the post content or block or styles that should be added to all "block preview"?

ntsekouras replied:
The fix should be in the preview and noticed this in this PR: https://github.com/WordPress/gutenberg/pull/35932/files#r735831633

--edit: I was only thinking of the appenders though.. I now noticed the placeholder styles here..

andrewserong replied:
I've moved these styles over to the live variation of the BlockPreview component. In addition to the placeholder styles, we also need the one for the buttons component so that Social Links buttons don't render greyed out.

So that this PR still tests well in isolation, I've left in the appender styling rule, but the approach to switch off the appender in #35932 looks like a better way to do it. I _think_ we'll still need the other styles, though.

youknowriad commented on packages/block-library/src/post-content/edit.js:
```diff
@@ -26,14 +28,25 @@ function ReadOnlyContent( { userCanEdit, postType, postId } ) {
 	) : (
 		<div { ...blockProps }>
-			<RawHTML key="html">{ content?.rendered }</RawHTML>
+			<BlockPreview
```
This is going to add an addition div container that might break alignments and things like that, I wonder if there's a way to render a preview without container or reuse the container above. cc @ntsekouras

andrewserong replied:
I've had a go at removing the wrapper div in BlockPreview by adding an `__experimentalAsButton` prop. It seems to work so far in testing, but let me know if you can think of a better approach.

youknowriad commented on packages/block-library/src/post-content/edit.js:
```diff
@@ -26,14 +28,25 @@ function ReadOnlyContent( { userCanEdit, postType, postId } ) {
-			<RawHTML key="html">{ content?.rendered }</RawHTML>
+			<BlockPreview
+				blocks={ blocks }
+				tabIndex={ -1 }
```
Why do we need the tabIndex prop?

andrewserong replied:
This one's pretty subtle, but the `tabIndex={ -1 }` is required because the live BlockPreview sets `tabIndex={ 0 }` by default, which makes the block preview itself reachable when you press the up and down keys to navigate between blocks. By adding the `-1` we remove its div from being reached by the up/down navigation. Here's a gif of what it looks like when we move between blocks if we remove this line:

<img src="https://user-images.githubusercontent.com/14988353/138984109-64c2c4ef-9488-4fae-8d51-b2caaf2c2239.gif" width="600" />

In the above screengrab:

1. The group block is selected
2. I press the UP arrow once, and the Post Content block is selected, but the preview within it is highlighted
3. I press the UP arrow again, and the Post Content block is still selected (instead of moving up to the next block), but the focus now moves to the parent container

andrewserong replied:
Update: I've added an `__experimentalAsButton` prop which will allow us to switch off the `div` wrapper that acted as a button, so I could remove the `tabIndex` prop.

andrewserong committed:
Remove tabIndex prop, replace with __experimentalAsButton, move CSS styles to block-preview

andrewserong committed:
Remove tabIndex prop from post-content edit

andrewserong commented:
Thanks for the review @youknowriad and the input @ntsekouras! I think I've managed to address the three pieces of feedback by doing the following:

* Added an `__experimentalAsButton` prop to the Block Preview, that defaults to `true`. By switching this to `false`, we can render the BlockPreview without a wrapper `div`, which will remove the additional layer of nesting.
* Moved the CSS rules introduced here to the BlockPreview component, and added an extra class for the live version of the component to use via the `Disabled` component.
* Since `__experimentalAsButton` removes the wrapper div that acted as a button, I've removed the `tabIndex` prop.

There's probably still some overlap between this PR and #35932, so let me know if the above changes need tweaking further (or if it looks like the approach for removing the BlockPreview wrapper isn't going to work). Thanks!

youknowriad commented on packages/block-editor/src/components/block-preview/live.js:
```diff
@@ -8,17 +8,26 @@ import { Disabled } from '@wordpress/components';
+	__experimentalAsButton = true,
+} ) {
+	const blockList = (
+		<Disabled className="block-editor-block-preview__live-content">
```
While this removes on container div, I believe `<BlockList>` also adds its own container which means we'd still break alignments rules if the post content block has a layout. right?

andrewserong replied:
Ah, yes, of course, I see now. I _think_ it might be possible for us to pass just the `.wp-container-` class down to the `BlockList` to get alignments working. I had a hack around today and managed to get it working, but ran into an issue with [withDataAlign](https://github.com/wordpress/gutenberg/blob/32ecd97e988174b6401289c44200135c224ca104/packages/block-editor/src/hooks/align.js#L172) not returning the `data-align` props for the Group blocks _within_ the block preview (e.g. within post content). I could hard-code it to always show, but I'll need to do a little more digging to get it working cleanly.

I'll keep exploring tomorrow!

youknowriad replied:
What about finding a way to use `BlockListItems` instead of `BlockList` kind of like `useInnerBlocksProps` does.

andrewserong replied:
Oh, that's a good idea, thanks! I'll have a play with that tomorrow 👍

andrewserong committed:
Add useDisabled hook, pass layout with alignments, use BlockListItems

andrewserong commented:
@youknowriad, I've managed to get this working in https://github.com/WordPress/gutenberg/pull/35863/commits/8b4c0d85e12b1b1fa039a2f7ec87d24031fc5d51 by removing the final two layers of div nesting. Firstly, `BlockListItems` worked well to get rid of one layer of `div`. The next layer to remove was the one introduced by the `Disabled` component.

I've done the following, but I'm sure this will need some cleaning up / re-organising to neaten it up. I'd love feedback on where best to place the `useDisabled` hook, if it looks suitable:

* Added a `useDisabled` hook that copies most of the logic from the `Disabled` component, but allows it to be used on an arbitrary component without adding a wrapper — this seemed important to do in order to be able to remove the additional `div` wrapper. But, I'm not sure where this hook would be best placed. Perhaps it can be stored in a shared location, and the `Disabled` component could be refactored to use this hook, too?
* Because of removing the additional layer of nesting, the classnames / styles need to be applied to the parent container, which is pretty awkward. I've added the classnames to the parent in this change, but it's pulling styles from another component, so I'm sure there's a neater way to do it, but I haven't found it just yet.
* Swapped `BlockItems` for `BlockListItems` which worked nicely to remove the extra `div`.
* Added in a hard-coded layout to pass to the `BlockListItems` in the `__experimentalLayout` prop. This is needed, because otherwise the root level components within the list of items aren't rendered with the `data-align` properties — I think this is because by default we treat a layout as if it's for the root level (where we don't use alignments), but in this kind of preview we _do_ want to render the alignments.

The end result is that I believe this PR is now working as we expect it to, with the Post Content block correctly rendering custom content widths in the editor:

![image](https://user-images.githubusercontent.com/14988353/139387922-f66d9eb2-5ae8-428f-96d4-f30da636fb5a.png)

I'm at the end of my week, and have only just pushed the code change, so I'm sure I've missed things in the implementation (I also haven't checked to see if this introduces regressions elsewhere). I'll be away on Monday/Tuesday next week, but can pick this up again on Wednesday. Of course, if it's needed to get this in sooner, feel free to commit in the PR if there's any urgency. Otherwise I'll dig in again first thing Wednesday.

Thanks! 🙇

youknowriad commented on packages/block-editor/src/components/block-preview/live.js:
```diff
@@ -6,19 +6,47 @@ import { Disabled } from '@wordpress/components';
+const layout = {
+	type: 'default',
+	alignments: [ ...DEFAULT_CONTROLS, ...WIDE_CONTROLS ],
+};
```
It seems the layout should be inherited from the parent block here, potentially it could be something completely different (like a flex block...)

andrewserong replied:
I've switched to inheriting the layout in the post-content block and passing it down to the new `useBlockPreview` hook

youknowriad commented on packages/block-editor/src/components/block-preview/live.js:
```diff
@@ -6,19 +6,47 @@ import { Disabled } from '@wordpress/components';
+} ) {
+	let blockList;
+
+	if ( __experimentalAsButton ) {
```
Is this prop still needed?

youknowriad replied:
Oh you changed the default value to `true` so I guess it's still needed for default previews

andrewserong replied:
I've removed the changes to `live.js` now that we're using a separate hook instead.

youknowriad commented on packages/block-library/src/post-content/use-disabled.js:
```diff
@@ -0,0 +1,92 @@
+	'TEXTAREA',
+];
+
+export function useDisabled() {
```
I think the right place for it might be `@wordpress/compose` WDYT?

andrewserong replied:
Good idea, I've moved it over.

youknowriad reviewed (changes requested):
Thanks for your work, I think we're going in the right direction but this PR turned out a lot more complex than I was expecting initially :) 

Reading the PR, it became clearer to me that what we really want here is a hook `useBlockPreview` that is very similar to `useInnerBlocksProps` hook but for previews (do not insert the blocks into the block store but relies on an alternative BlockEditorProvider) and the `BlockPreview` component would just rely on this hook internally

I think this way we won't have any issues with the classes needed to be manually applied to the wrapper... and no need for extra isButton prop...

andrewserong commented:
Thanks again for the helpful feedback, Riad, much appreciated! 🙇 

> I think we're going in the right direction but this PR turned out a lot more complex than I was expecting initially :)

Funny how these things go! All it takes is a couple of little edge cases sometimes, and then the scope creeps up on us 😀

> Reading the PR, it became clearer to me that what we really want here is a hook `useBlockPreview` that is very similar to `useInnerBlocksProps` hook but for previews

That's a terrific idea, that's just the kind of abstraction I was hoping we'd hone in on, thanks for the suggestion! I'll dig into that next week, hopefully it should make things much cleaner. Cheers!

andrewserong committed:
Refactor block preview behaviour to a useBlockPreview hook for the post content block

andrewserong committed:
Remove useDisabled hook from post content now that it has been moved to the compose package

andrewserong commented:
Thanks again @youknowriad, I've moved the BlockPreview behaviour for the Post Content block to a separate `useBlockPreview` hook, which in turn uses the `useDisabled` hook, too. I _think_ it's working pretty well (and feels much neater).

It also wasn't hard to pass the `layout` settings down from the Post Content block to the `useBlockPreview` hook so that's resolved the issue of rendering layouts: 👍 

![image](https://user-images.githubusercontent.com/14988353/140017798-880a8f02-36fc-4504-9fea-7f84ce9fba83.png)

I'm sure this'll need some more cleaning / tidying and testing, but since I'm at the end of my day, I've pushed where I'm up to. Let me know if there are any other immediate changes you think we should make. Otherwise, tomorrow I'll take another look with fresh eyes and see what I've missed / what else needs testing.

I thought I'd leave the refactoring of the existing BlockPreview component to use the `useBlockPreview` hook internally to a follow-up PR as it's not strictly speaking necessary for this bug fix, so I imagine that might be a good one for me to look into after we've shipped the next Gutenberg version. Let me know if you think we should do that first, though, and I can take a look. Cheers!

youknowriad reviewed (approved):
This is looking good to me. I'd appreciate another review potentially. 

cc @ntsekouras @jasmussen

jasmussen commented:
Thanks for the ping. I'm not entirely sure what to test for, so just taking the site editor for a quick spin seemed to work as intended. In the TT2 templates, the post content block is set to inherit styles by default, but tweaking those also worked as they should.

gziolo commented on packages/block-library/src/post-content/edit.js:
```diff
@@ -13,28 +14,45 @@ import {
+
+	const blockPreviewProps = useBlockPreview( blockProps, {
+		blocks,
+		__experimentalLayout: themeSupportsLayout ? usedLayout : undefined,
```
Is the logic for the layout something that could be moved to `useBlockPreview`? You could potentially read the `layout` attribute from the block edit context.

gziolo replied:
In theory, `useBlockProps` can be called inside `useBlockPreview`, too. We might need the same hook for Comment Content block one day so I'm thinking about how to make this hook easier to use in other places.

andrewserong replied:
That's a neat idea — I think for the moment, I'm leaning toward keeping this particular layout logic outside of the `useBlockPreview` hook, in case it's possible for us to use the hook in other circumstances where `useBlockProps` isn't appropriate. I'm thinking of arbitrary block previews that are used elsewhere in the repo, for example in the block switcher or for block patterns. They currently use the `BlockPreview` component directly, but it could be good to retain some flexibility for the hook in case we wanted to roll it out there, (or internally in the `BlockPreview` component itself?

gziolo replied:
You are absolutely right. I didn't think about the original use case of `<BlockPreview />` that was hidden in the changes for this PR. I guess the usage of the preview inside the `edit` is rather an edge case so we shouldn't expect that the block edit context is set here :+1:

gziolo commented on packages/block-editor/src/components/block-preview/index.js:
```diff
@@ -63,3 +69,57 @@ export function BlockPreview( {
+ * returns. Optionally, you can also pass any other props through this hook, and
+ * they will be merged and returned.
+ *
+ * @param {Object}    props                        Optional. Props to pass to the element. Must contain
```
Given that `props` are optional, they can be moved to options to remove the need to pass an empty object when someone wants only to set options.

andrewserong replied:
Good suggestion, thanks — I've moved `props` to be inside `options` so this function now only has one param 👍

ntsekouras commented on packages/compose/src/hooks/use-disabled/index.js:
```diff
@@ -0,0 +1,92 @@
+	'TEXTAREA',
+];
+
+export default function useDisabled() {
```
We could use this hook in `Disabled` component. I think it's okay for a follow up though.

andrewserong replied:
Yes, if this PR lands, I'm happy to look into that as a follow-up to make it a bit more DRY.

ntsekouras commented on packages/block-library/src/post-content/edit.js:
```diff
@@ -13,28 +14,45 @@ import {
 	);
 	const blockProps = useBlockProps();
+
+	const themeSupportsLayout = useSelect( ( select ) => {
```
Since now both `EditableContent` and `ReadOnlyContent` use the `layout` we can remove this logic from both and add it only to `Content`.

andrewserong replied:
Thanks, I've moved the layout logic up to the Content component 👍

ntsekouras commented on packages/block-library/src/post-content/edit.js:
```diff
@@ -13,28 +14,45 @@ import {
+	const defaultLayout = useSetting( 'layout' ) || {};
+	const usedLayout = !! layout && layout.inherit ? defaultLayout : layout;
+
+	const rawContent = content?.raw;
```
I'm little hesitant with this change from `content.rendered` to `content.raw`. I think that this might lead to security issues with private content.. @peterwilsoncc can you share your thoughts?

peterwilsoncc replied:
That's correct: previews need to use the rendered version as the raw version may not be available to all users. For example authors don't have raw rights for others' posts, contributors for any published post.

andrewserong replied:
Thanks for raising this @ntsekouras and for the feedback @peterwilsoncc! In the use case in this PR, unfortunately I don't think it'll work to use `content.rendered` as we need the raw markup of blocks in order to be able to parse them and have the styles be rendered properly by the preview.

As a bit of background, this PR is an alternative / workaround for the issue of rendering block supports styles on the server in order to display styles correctly in the editor. We currently have no way of retrieving the block supports styles from the server via an API request, so the next best option in this PR was to allow the editor to generate the styles from the block markup.

In the case of the Post Content block, in most cases, I imagine that users will see this block in the editor when editing template parts rather than in an isolated post. Because author and contributor roles don't have access to the site editor, in practice, I'm wondering if it isn't _too_ much of an issue if we have the Post Content preview depend on access to `content.raw`. We already only render the blocks if `content.raw` is available, otherwise this falls back to rendering an empty preview, so on balance I think it could be more beneficial to prioritise the correct rendering for the admin view of the block, rather than focusing on author/contributor roles being able to view it in the editor.

ntsekouras reviewed (commented):
Thanks for your work here @andrewserong !

My main concern is the security related one with `content.raw`. We need to be sure not to expose data where we shouldn't be.

andrewserong committed:
Combine props into options param of useBlockPreview

andrewserong committed:
Combine layout logic

andrewserong commented:
Thanks for the feedback, everyone! 🙇  I've done the following small changes:

* Moved the `props` param of the `useBlockPreview` hook to be inside the `options` param so that the hook only has one required parameter.
* Moved the layout logic in the Post Content edit component up to the `Content` component, to remove some duplication in `EditableContent` and `ReadOnlyContent`.

I think the last remaining question is whether or not we can comfortably use `content.raw`. From the sounds of it, if it's a matter of whether we are _able_ to show a preview to non-admin users, I personally think that feature loss is probably worth it, since the Post Content block is most likely to be used in contexts where admins will be editing (e.g. in the site editor) instead of author or contributor roles, who will just be in the post editor. I'm not _too_ familiar with the inner workings of `content.raw` though, so please do let me know if you think it'll be a blocker. This PR relies on the approach of using raw block markup in order to generate the preview, so if using `content.raw` isn't viable, it will probably be a blocker for this PR progressing. I noticed that the [Latest Posts block](https://github.com/wordpress/gutenberg/blob/63f390f53eddc6a1c6ec85f5e29e40fe04e3a258/packages/block-library/src/latest-posts/edit.js#L546), [Navigation block](https://github.com/wordpress/gutenberg/blob/4ea694f047959041de026cd11e8e648c5dae7b5f/packages/block-library/src/navigation/menu-items-to-blocks.js#L41), and [Template parts](https://github.com/wordpress/gutenberg/blob/bb0c0476c49e81ce56ebd241f1d744fe400aa6a4/packages/block-library/src/template-part/edit/selection/template-part-previews.js#L59) also use `content.raw` within editor previews, so there's at least some precedent there, if we choose to proceed with the `content.raw` approach.

ntsekouras commented:
Thanks for all your work here @andrewserong !

>This PR relies on the approach of using raw block markup in order to generate the preview, so if using content.raw isn't viable, it will probably be a blocker for this PR progressing

I think this is the case here unfortunately and we should try to find some other way to resolve this issue.

peterwilsoncc commented:
Per discussion above about this PR being blocked, closing it off so an alternative approach can be considered.
